Code/src/main/java/nl/utwente/group10/haskell/expr/Function.java

```diff
@@ -0,0 +1,167 @@
+package nl.utwente.group10.haskell.expr;
+
+import nl.utwente.group10.haskell.env.Env;
+import nl.utwente.group10.haskell.exceptions.HaskellException;
+import nl.utwente.group10.haskell.hindley.GenSet;
+import nl.utwente.group10.haskell.hindley.HindleyMilner;
+import nl.utwente.group10.haskell.type.FuncT;
+import nl.utwente.group10.haskell.type.Type;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Represents a user-defined Haskell function. Holds arguments which can be used for building the function.
+ */
+public class Function extends Expr {
+
+    /**
+     * An argument for a function. Can be used as a variable in a function definition.
+     */
+    public class FunctionArgument extends Expr {
+
+        /** The function this argument is for. */
+        private Function function;
+
+        /**
+         * @param function The function this argument is for.
+         */
+        protected FunctionArgument(Function function) {
+            this.function = function;
+        }
+
+        /**
+         * @return The 0-indexed position of this argument in the function definition.
+         */
+        public int getPosition() {
+            return this.function.arguments.indexOf(this);
+        }
+
+        @Override
+        public Type analyze(Env env, GenSet genSet) throws HaskellException {
+            return HindleyMilner.makeVariable(); // TODO Make sure that this is correctly unified and/or make this editable.
+        }
+
+        @Override
+        public String toHaskell() {
+            return String.format("%s%d", Function.ARG_PREFIX, this.getPosition());
+        }
+
+        @Override
+        public String toString() {
+            return String.format("%s%d", Function.ARG_PREFIX, this.getPosition());
+        }
+    }
+
+    /** The prefix for function argument names. */
+    protected static final String ARG_PREFIX = "arg";
+
+    /** The arguments for this function. */
+    private List<FunctionArgument> arguments;
+
+    /** The expression that forms the base of this function. */
+    private Optional<Expr> expr;
+
+    /**
+     * @param expr The expression that forms the base of this function.
+     */
+    public Function(Expr expr) {
+        this.arguments = new ArrayList<>();
+        this.expr = Optional.ofNullable(expr);
+    }
+
+    /**
+     */
+    public Function() {
+        this(null);
+    }
```

Comment:
I'm not sure if it would work, but I'd rather see an immutable Function representation that's lightweight enough to build anew every time we do `asExpr`.

It seems that this Function is designed to be mutable and have its state changed when needed, even starting out in a kinda-sorta valid state. All the other subclasses of Expr are pretty much ready to roll and immutable once they're created (with the exception of Value's `setValue`, perhaps from an oversight). So I'd say, make this one immutable as well:

- Drop no-arg constructor: have to know expr *and number of args* before turning them into a Function.
- No setters or list manipulation on arguments.
- Make some things `final` for good measure, like in Apply.
- No real need for `getArguments()` I think - a `getArgument(i)` would perhaps be useful though.